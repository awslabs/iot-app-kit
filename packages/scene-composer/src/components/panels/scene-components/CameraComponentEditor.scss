.tm-camera-sub-form-label {
  padding: 10px 0px 0px 0px;
}
