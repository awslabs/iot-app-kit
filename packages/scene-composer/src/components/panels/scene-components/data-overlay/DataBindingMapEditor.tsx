import React, { useCallback } from 'react';
import { Box, Button, SpaceBetween } from '@awsui/components-react';
import { useIntl } from 'react-intl';
import { isEmpty } from 'lodash';
import styled from 'styled-components';

import { IValueDataBinding, IValueDataBindingProvider } from '../../../../interfaces';
import { IDataOverlayComponentInternal } from '../../../../store';
import { ValueDataBindingBuilder } from '../ValueDataBindingBuilder';
import { generateUUID } from '../../../../utils/mathUtils';
import { Divider } from '../../../Divider';

import { DataBindingMapNameEditor } from './DataBindingMapNameEditor';

interface IDataBindingMapEditorProps {
  valueDataBindingProvider: IValueDataBindingProvider | undefined;
  component: IDataOverlayComponentInternal;
  onUpdateCallback: (componentPartial: Partial<IDataOverlayComponentInternal>, replace?: boolean | undefined) => void;
}

const RemoveButtonContainer = styled.div`
  float: right;
  height: 12px;
`;
const Spacing = styled.div`
  height: 8px;
`;

export const DataBindingMapEditor: React.FC<IDataBindingMapEditorProps> = ({
  valueDataBindingProvider,
  component,
  onUpdateCallback,
}) => {
  const intl = useIntl();

  const onBindingChange = useCallback(
    (valueDataBinding: IValueDataBinding, index) => {
      // we don't want to merge the dataBindingContext, so we'll need to manually replace it
      const updatedComponent = {
        ...component,
        valueDataBindings: [...component.valueDataBindings],
      };
      updatedComponent.valueDataBindings[index] = {
        ...updatedComponent.valueDataBindings[index],
        valueDataBinding,
      };
      onUpdateCallback(updatedComponent, true);
    },
    [component.valueDataBindings, onUpdateCallback],
  );

  const onRemoveBinding = useCallback(
    (index) => {
      const newBindings = component.valueDataBindings.filter((_, i) => i !== index);
      onUpdateCallback({ ...component, valueDataBindings: newBindings }, true);
    },
    [component.valueDataBindings, onUpdateCallback],
  );

  const onAddBinding = useCallback(() => {
    const newBindings = [...component.valueDataBindings];
    newBindings.push({ bindingName: generateUUID() });
    onUpdateCallback({ valueDataBindings: newBindings });
  }, [component.valueDataBindings, onUpdateCallback]);

  return (
    <SpaceBetween size='s'>
      {valueDataBindingProvider && (
        <>
          {!isEmpty(component.valueDataBindings) &&
            component.valueDataBindings.map(({ bindingName, valueDataBinding }, index) => (
              <Box key={index}>
                <Divider />

                <RemoveButtonContainer>
                  <Button
                    data-test-id='remove-binding-button'
                    iconName='close'
                    variant='icon'
                    iconAlign='right'
                    onClick={() => onRemoveBinding(index)}
                  />
                </RemoveButtonContainer>
                <Spacing />

                <DataBindingMapNameEditor
                  bindingName={bindingName}
                  index={index}
                  valueDataBindings={component.valueDataBindings}
                  onUpdateCallback={onUpdateCallback}
                />

                <Box margin={{ vertical: 's' }}>
                  <ValueDataBindingBuilder
                    componentRef={component.ref}
                    binding={valueDataBinding}
                    valueDataBindingProvider={valueDataBindingProvider}
                    onChange={(v) => onBindingChange(v, index)}
                  />
                </Box>
              </Box>
            ))}
          <Button data-test-id='add-binding-button' onClick={onAddBinding}>
            {intl.formatMessage({ defaultMessage: 'Add data binding', description: 'Button text' })}
          </Button>
        </>
      )}
    </SpaceBetween>
  );
};
