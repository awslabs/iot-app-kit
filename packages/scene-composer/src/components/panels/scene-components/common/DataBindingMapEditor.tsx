import { Box, Button, SpaceBetween } from '@awsui/components-react';
import { isEmpty } from 'lodash';
import React, { useCallback } from 'react';
import { useIntl } from 'react-intl';
import styled from 'styled-components';

import { IValueDataBinding, IValueDataBindingProvider } from '../../../../interfaces';
import { Component } from '../../../../models/SceneModels';
import { ISceneComponentInternal } from '../../../../store';
import { Divider } from '../../../Divider';
import { DataBindingMapNameEditor } from '../data-overlay/DataBindingMapNameEditor';

import { ValueDataBindingBuilder } from './ValueDataBindingBuilder';

export interface ComponentWithDataBindings extends ISceneComponentInternal {
  valueDataBindings: Component.IDataBindingMap[] | Component.ValueDataBindingNamedMap[];
}

interface IDataBindingMapEditorProps {
  hasBindingName: boolean;
  valueDataBindingProvider: IValueDataBindingProvider | undefined;
  component: ComponentWithDataBindings;
  onUpdateCallback: (componentPartial: Partial<ComponentWithDataBindings>, replace?: boolean | undefined) => void;
  numFields?: number;
  hasRemoveButton?: boolean;
  skipFirstDivider?: boolean;
  allowPartialBinding?: boolean;
}

const RemoveButtonContainer = styled.div`
  float: right;
  height: 12px;
`;
const Spacing = styled.div`
  height: 8px;
`;

export const DataBindingMapEditor: React.FC<IDataBindingMapEditorProps> = ({
  hasBindingName,
  valueDataBindingProvider,
  component,
  numFields,
  hasRemoveButton,
  onUpdateCallback,
  skipFirstDivider,
  allowPartialBinding,
}) => {
  const intl = useIntl();

  const onBindingChange = useCallback(
    (valueDataBinding: IValueDataBinding, index) => {
      // we don't want to merge the dataBindingContext, so we'll need to manually replace it
      const updatedComponent = {
        ...component,
        valueDataBindings: [...component.valueDataBindings],
      };
      updatedComponent.valueDataBindings[index] = {
        ...updatedComponent.valueDataBindings[index],
        valueDataBinding,
      };
      onUpdateCallback(updatedComponent, true);
    },
    [component.valueDataBindings, onUpdateCallback],
  );

  const onRemoveBinding = useCallback(
    (index) => {
      const newBindings = component.valueDataBindings.filter((_, i) => i !== index);
      onUpdateCallback({ ...component, valueDataBindings: newBindings }, true);
    },
    [component.valueDataBindings, onUpdateCallback],
  );

  return (
    <SpaceBetween size='s'>
      {valueDataBindingProvider && (
        <>
          {!isEmpty(component.valueDataBindings) &&
            component.valueDataBindings.map((binding, index) => (
              <Box key={index}>
                {(index > 0 || !skipFirstDivider) && <Divider />}

                {hasRemoveButton && (
                  <RemoveButtonContainer>
                    <Button
                      ariaLabel={intl.formatMessage({
                        defaultMessage: 'Remove data binding',
                        description: 'Button label',
                      })}
                      data-testid='remove-binding-button'
                      iconName='close'
                      variant='icon'
                      iconAlign='right'
                      onClick={() => onRemoveBinding(index)}
                    />
                  </RemoveButtonContainer>
                )}
                <Spacing />
                {hasBindingName && (
                  <DataBindingMapNameEditor
                    bindingName={(binding as Component.ValueDataBindingNamedMap).bindingName}
                    index={index}
                    valueDataBindings={component.valueDataBindings as Component.ValueDataBindingNamedMap[]}
                    onUpdateCallback={onUpdateCallback}
                  />
                )}

                <Box margin={{ vertical: 's' }}>
                  <ValueDataBindingBuilder
                    allowPartialBinding={allowPartialBinding}
                    componentRef={component.ref}
                    binding={binding.valueDataBinding}
                    numFields={numFields}
                    valueDataBindingProvider={valueDataBindingProvider}
                    onChange={(v) => onBindingChange(v, index)}
                  />
                </Box>
              </Box>
            ))}
        </>
      )}
    </SpaceBetween>
  );
};
