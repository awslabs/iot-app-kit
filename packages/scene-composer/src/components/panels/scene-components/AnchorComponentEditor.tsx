import { AttributeEditor, FormField, Grid, Input, Select, SpaceBetween, TextContent } from '@awsui/components-react';
import { IconLookup, findIconDefinition } from '@fortawesome/fontawesome-svg-core';
import { debounce } from 'lodash';
import React, { useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react';
import { defineMessages, useIntl } from 'react-intl';

import { getGlobalSettings } from '../../../common/GlobalSettings';
import { SCENE_ICONS } from '../../../common/constants';
import { sceneComposerIdContext } from '../../../common/sceneComposerIdContext';
import { COMPOSER_FEATURES, DefaultAnchorStatus, IValueDataBinding, SceneResourceType } from '../../../interfaces';
import { IAnchorComponentInternal, ISceneComponentInternal, useSceneDocument, useStore } from '../../../store';
import { shallowEqualsArray } from '../../../utils/objectUtils';
import { i18nSceneIconsKeysStrings } from '../../../utils/polarisUtils';
import { convertToIotTwinMakerNamespace, getSceneResourceInfo } from '../../../utils/sceneResourceUtils';
import { colors } from '../../../utils/styleUtils';
import { IComponentEditorProps } from '../ComponentEditor';
import { TextInput } from '../CommonPanelComponents';
import useDynamicScene from '../../../hooks/useDynamicScene';

import { ValueDataBindingBuilder } from './common/ValueDataBindingBuilder';
import { ColorSelectorCombo } from './tag-style/ColorSelectorCombo/ColorSelectorCombo';
import { DecodeSvgString } from './tag-style/ColorSelectorCombo/ColorSelectorComboUtils/DecodeSvgString';
import { IconPicker } from './tag-style/IconPicker/IconPicker';

export const convertParamsToKeyValuePairs = (params: Record<string, string>) => {
  return Object.keys(params).map((key) => {
    return { key, value: params[key] };
  });
};

const i18nIconStrings = defineMessages({
  [DefaultAnchorStatus.Info]: { defaultMessage: 'Info icon', description: 'Icon name label' },
  [DefaultAnchorStatus.Warning]: { defaultMessage: 'Warning icon', description: 'Icon name label' },
  [DefaultAnchorStatus.Error]: { defaultMessage: 'Error icon', description: 'Icon name label' },
  [DefaultAnchorStatus.Video]: { defaultMessage: 'Video icon', description: 'Icon name label' },
  [DefaultAnchorStatus.Custom]: { defaultMessage: 'Custom icon', description: 'Icon name label' },
});

export type IAnchorComponentEditorProps = IComponentEditorProps;

// TODO: This component is very poorly written, has too many responsibilities, and is using questionable hacks (like updating state on useEffect, and deouncing it to prevent a rerender cycle loop)
export const AnchorComponentEditor: React.FC<IAnchorComponentEditorProps> = ({
  node,
  component,
}: IAnchorComponentEditorProps) => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const tagStyle = getGlobalSettings().featureConfig[COMPOSER_FEATURES.TagStyle];
  const updateComponentInternal = useStore(sceneComposerId)((state) => state.updateComponentInternal);
  const valueDataBindingProvider = useStore(sceneComposerId)(
    (state) => state.getEditorConfig().valueDataBindingProvider,
  );
  const anchorComponent = component as IAnchorComponentInternal;
  const [items, setItems] = useState<{ key: string; value: string; constraintText?: string }[]>([]);
  const hasDuplicateKeyRef = useRef<boolean>(false);
  const { listSceneRuleMapIds } = useSceneDocument(sceneComposerId);
  const intl = useIntl();

  const dynamicSceneEnabled = useDynamicScene();

  const ruleMapIds = listSceneRuleMapIds();
  const selectedRuleMapId =
    anchorComponent.ruleBasedMapId && ruleMapIds.includes(anchorComponent.ruleBasedMapId)
      ? anchorComponent.ruleBasedMapId
      : null;

  const onUpdateCallbackDebounced = useCallback(
    debounce(
      (componentPartial: any, replace?: boolean) => {
        const componentPartialWithRef: ISceneComponentInternal = { ref: component.ref, ...componentPartial };
        updateComponentInternal(node.ref, componentPartialWithRef, replace);
      },
      dynamicSceneEnabled ? 1000 : 100,
    ), // TODO: Temporary solution for the error when updating entity too frequent. Will implement a better solution for GA.
    [node.ref, component.ref],
  );

  const onUpdateCallbackForChosenColor = useCallback(
    debounce((componentPartial: any, replace?: boolean) => {
      const componentPartialWithRef: ISceneComponentInternal = { ref: component.ref, ...componentPartial };
      updateComponentInternal(node.ref, componentPartialWithRef, replace);
    }, 300),
    [node.ref, component.ref],
  );

  const onUpdateCallback = useCallback(
    (componentPartial: any, replace?: boolean) => {
      const componentPartialWithRef: ISceneComponentInternal = { ref: component.ref, ...componentPartial };
      updateComponentInternal(node.ref, componentPartialWithRef, replace);
    },
    [node.ref, component.ref],
  );

  useMemo(() => {
    setItems(convertParamsToKeyValuePairs(anchorComponent.navLink?.params || {}));
  }, [anchorComponent]);

  useEffect(() => {
    // update the state if the items are changed
    const params = convertParamsToKeyValuePairs(anchorComponent.navLink?.params || {});
    if (!shallowEqualsArray(items, params)) {
      const updatedComponent = { ...anchorComponent } as IAnchorComponentInternal;
      updatedComponent.navLink = Object.assign({}, updatedComponent.navLink);
      if (!hasDuplicateKeyRef.current) {
        updatedComponent.navLink.params = Object.assign({}, Object.fromEntries(items.map((i) => [i.key, i.value])));
        onUpdateCallbackDebounced(updatedComponent, true);
      }
    }
  }, [items]);

  const updateItem = (itemIndex: number, newKey: string, newValue: string) => {
    const updatedItems = [...items];
    // checking duplicate key
    const keySet = new Set<string>();
    let keyConstraintText = '';
    hasDuplicateKeyRef.current = false;
    for (const item of items) {
      if (keySet.has(newKey)) {
        keyConstraintText = 'The key already exists.';
        hasDuplicateKeyRef.current = true;
        break;
      } else {
        keySet.add(item.key);
      }
    }

    updatedItems[itemIndex] = { key: newKey, value: newValue, constraintText: keyConstraintText };
    setItems(updatedItems);
  };

  const addItem = (key: string, value: string) => {
    const updatedItems = [...items, { key, value }];
    setItems(updatedItems);
  };

  const deleteItem = (itemIndex: number) => {
    const updatedItems = [...items];
    updatedItems.splice(itemIndex, 1);
    setItems(updatedItems);
  };

  const onUpdateNavLinkParams = useCallback(
    (index, prop: 'key' | 'value', value) => {
      const newKey = prop === 'key' ? value : items[index].key;
      const newValue = prop === 'value' ? value : items[index].value;
      updateItem(index, newKey, newValue);
    },
    [items],
  );

  const iconOptions = useMemo(() => {
    return Object.keys(SCENE_ICONS)
      .filter((icon) => (!tagStyle && icon !== 'Custom') || tagStyle)
      .map((sceneIcon) => {
        return {
          label: intl.formatMessage(i18nSceneIconsKeysStrings[sceneIcon]) || sceneIcon,
          value: sceneIcon,
        };
      });
  }, [tagStyle]);

  const iconSelectedOptionIndex = useMemo(() => {
    if (!anchorComponent.icon) {
      return -1;
    }

    const sceneResourceInfo = getSceneResourceInfo(anchorComponent.icon);
    return Object.keys(SCENE_ICONS).indexOf(sceneResourceInfo.value);
  }, [anchorComponent, iconOptions]);

  const iconString = useMemo(() => {
    const iconStrings = Object.values(SCENE_ICONS);
    if (iconSelectedOptionIndex < 0 || iconSelectedOptionIndex >= iconStrings.length) {
      return;
    }

    return btoa(iconStrings[iconSelectedOptionIndex]);
  }, [iconSelectedOptionIndex]);

  const ruleOptions = ruleMapIds
    .concat(
      selectedRuleMapId
        ? intl.formatMessage({
            defaultMessage: 'No Rule',
            description: 'signify No rule option to be selected in a drop down menu',
          })
        : [],
    )
    .map((ruleMapId) => ({ label: ruleMapId, value: ruleMapId }));

  const hasIcon = iconSelectedOptionIndex >= 0;
  const iconGridDefinition = hasIcon ? [{ colspan: 10 }, { colspan: 2 }] : [{ colspan: 12 }];
  const isCustomStyle = tagStyle && iconOptions[iconSelectedOptionIndex]?.value === 'Custom';
  const customIcon = (anchorComponent.customIcon ?? { prefix: 'fas', iconName: 'info' }) as IconLookup;

  return (
    <SpaceBetween size='s'>
      <FormField label={intl.formatMessage({ defaultMessage: 'Tag style', description: 'Form field label' })}>
        <Grid gridDefinition={iconGridDefinition}>
          <Select
            data-testid='anchor-default-icon-select'
            selectedOption={hasIcon ? iconOptions[iconSelectedOptionIndex] : null}
            onChange={(e) => {
              if (e.detail.selectedOption.value) {
                const icon = convertToIotTwinMakerNamespace(SceneResourceType.Icon, e.detail.selectedOption.value);
                onUpdateCallback({
                  icon,
                });
              }
            }}
            options={iconOptions}
            selectedAriaLabel={intl.formatMessage({
              defaultMessage: 'Selected',
              description:
                'Specifies the localized string that describes an option as being selected. This is required to provide a good screen reader experience',
            })}
            placeholder={intl.formatMessage({ defaultMessage: 'Choose an icon', description: 'placeholder' })}
          />
          {hasIcon &&
            (isCustomStyle ? (
              <DecodeSvgString
                selectedColor={anchorComponent.chosenColor ?? colors.infoBlue}
                iconString={iconString!}
                customIcon={findIconDefinition(customIcon)}
                width='32px'
                height='32px'
                ariaLabel={intl.formatMessage(i18nIconStrings[iconOptions[iconSelectedOptionIndex]?.value])}
              />
            ) : (
              <img
                aria-label={intl.formatMessage(i18nIconStrings[iconOptions[iconSelectedOptionIndex]?.value])}
                width='32px'
                height='32px'
                src={`data:image/svg+xml;base64,${iconString}`}
              />
            ))}
        </Grid>
      </FormField>
      {isCustomStyle && (
        <FormField>
          <SpaceBetween size='m'>
            <ColorSelectorCombo
              color={anchorComponent.chosenColor ?? colors.customBlue}
              onSelectColor={(pickedColor) => {
                onUpdateCallbackForChosenColor({
                  chosenColor: pickedColor,
                });
              }}
              onUpdateCustomColors={(chosenCustomColors) =>
                onUpdateCallbackForChosenColor({ customColors: chosenCustomColors })
              }
              customColors={anchorComponent.customColors}
              colorPickerLabel={intl.formatMessage({ defaultMessage: 'Color', description: 'Color' })}
              customColorLabel={intl.formatMessage({ defaultMessage: 'Custom colors', description: 'Custom colors' })}
            />
            <IconPicker
              onSelectIconChange={(pickedIcon) => {
                onUpdateCallback({
                  customIcon: pickedIcon,
                });
              }}
              selectedIcon={customIcon}
              iconPickerLabel={intl.formatMessage({ defaultMessage: 'Icon', description: 'Icon' })}
              iconFilterText={intl.formatMessage({ defaultMessage: 'Find icons', description: 'Find icons' })}
              iconFilterTextAriaLabel={intl.formatMessage({
                defaultMessage: 'Filter icons',
                description: 'Filter icons',
              })}
            />
          </SpaceBetween>
        </FormField>
      )}

      {valueDataBindingProvider && (
        <ValueDataBindingBuilder
          componentRef={anchorComponent.ref}
          binding={anchorComponent.valueDataBinding}
          valueDataBindingProvider={valueDataBindingProvider}
          onChange={(valueDataBinding: IValueDataBinding) => {
            // we don't want to merge the dataBindingContext, so we'll need to manually replace it
            const updatedComponent = { ...component, valueDataBinding };
            onUpdateCallback(updatedComponent, true);
          }}
        />
      )}

      <FormField label={intl.formatMessage({ defaultMessage: 'Rule Id', description: 'Form field label' })}>
        <Select
          data-testid='anchor-rule-id-select'
          selectedOption={selectedRuleMapId ? { label: selectedRuleMapId, value: selectedRuleMapId } : null}
          onChange={(e) => {
            const ruleMapId = e.detail.selectedOption.value;
            if (ruleMapId) {
              if (ruleMapIds.includes(ruleMapId)) {
                onUpdateCallbackDebounced({ ruleBasedMapId: ruleMapId });
              } else {
                onUpdateCallbackDebounced({ ruleBasedMapId: undefined });
              }
            }
          }}
          options={ruleOptions}
          selectedAriaLabel={intl.formatMessage({
            defaultMessage: 'Selected',
            description:
              'Specifies the localized string that describes an option as being selected. This is required to provide a good screen reader experience',
          })}
          disabled={ruleMapIds.length === 0}
          placeholder={intl.formatMessage({ defaultMessage: 'Choose a rule', description: 'placeholder' })}
        />
      </FormField>

      <FormField label={intl.formatMessage({ defaultMessage: 'Link Target', description: 'Form field label' })}>
        <TextInput
          data-testid='anchor-link-target-input'
          value={anchorComponent.navLink?.destination || ''}
          setValue={(e) => onUpdateCallback({ navLink: { destination: e } })}
        />
      </FormField>

      <TextContent>
        {intl.formatMessage({ defaultMessage: 'Link Parameters', description: 'Form section title' })}
      </TextContent>
      <AttributeEditor
        data-testid='anchor-attribute-editor-select'
        items={items}
        isItemRemovable={() => true}
        definition={[
          {
            label: intl.formatMessage({ defaultMessage: 'Key', description: 'Field label' }),
            control: (item, index) => (
              <Input
                value={item.key}
                placeholder={intl.formatMessage({ defaultMessage: 'Key', description: 'placeholder' })}
                onChange={({ detail }) => {
                  onUpdateNavLinkParams(index, 'key', detail.value);
                }}
              />
            ),
            constraintText: (item) => item.constraintText,
          },
          {
            label: intl.formatMessage({ defaultMessage: 'Value', description: 'Field label' }),
            control: (item, index) => (
              <Input
                value={item.value}
                placeholder={intl.formatMessage({ defaultMessage: 'Value', description: 'placeholder' })}
                onChange={({ detail }) => {
                  onUpdateNavLinkParams(index, 'value', detail.value);
                }}
              />
            ),
          },
        ]}
        addButtonText={intl.formatMessage({
          defaultMessage: 'Add parameter',
          description: 'Add parameter Button Text',
        })}
        removeButtonText={intl.formatMessage({
          defaultMessage: 'Remove parameter',
          description: 'Remove parameter Button Text',
        })}
        empty={intl.formatMessage({
          defaultMessage: 'No parameter',
          description: 'visual text to show there are no parameters',
        })}
        onAddButtonClick={() => addItem('', '')}
        onRemoveButtonClick={({ detail: { itemIndex } }) => {
          deleteItem(itemIndex);
        }}
      />
    </SpaceBetween>
  );
};
