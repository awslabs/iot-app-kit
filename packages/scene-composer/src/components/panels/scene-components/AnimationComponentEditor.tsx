import React, { useCallback, useContext } from 'react';

import { accessStore, ISceneNodeInternal, IAnimationComponentInternal } from '../../../store';
import { sceneComposerIdContext } from '../../../common/sceneComposerIdContext';

import { AnimationViewStateEditor } from './AnimationViewStateEditor';

export type AnimationEditorProps = {
  node: ISceneNodeInternal;
  component: IAnimationComponentInternal;
};

export const AnimationComponentEditor: React.FC<AnimationEditorProps> = ({ ...props }) => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const updateComponentInternal = accessStore(sceneComposerId)((state) => state.updateComponentInternal);
  // istanbul ignore next
  const onUpdate = useCallback(
    (currentAnimations, noderef, component) => {
      updateComponentInternal(noderef, { ...component, currentAnimations });
    },
    [updateComponentInternal],
  );

  return <AnimationViewStateEditor {...props} onUpdate={onUpdate} />;
};
