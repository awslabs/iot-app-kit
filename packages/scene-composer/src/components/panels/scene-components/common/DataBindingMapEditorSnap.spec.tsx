import React from 'react';
import { render } from '@testing-library/react';

import { mockProvider } from '../../../../../tests/components/panels/scene-components/MockComponents';
import { IDataOverlayComponentInternal } from '../../../../store';

import { DataBindingMapEditor } from './DataBindingMapEditor';

describe('DataBindingMapEditor', () => {
  const component = {
    valueDataBindings: [
      {
        bindingName: 'binding-1',
        valueDataBinding: {
          dataBindingContext: 'random-1',
        },
      },
      {
        bindingName: 'binding-2',
        valueDataBinding: {
          dataBindingContext: 'random-2',
        },
      },
    ],
  } as unknown as IDataOverlayComponentInternal;
  const onUpdateCallbackMock = jest.fn();

  it('should render existing maps with bindings', async () => {
    const { container, queryAllByTestId } = render(
      <DataBindingMapEditor
        hasBindingName
        valueDataBindingProvider={mockProvider}
        component={component}
        onUpdateCallback={onUpdateCallbackMock}
      />,
    );

    expect(queryAllByTestId('binding-name-input').length).toBe(2);
    expect(container).toMatchSnapshot();
  });

  it('should render with no bindings', async () => {
    const { container, queryAllByTestId } = render(
      <DataBindingMapEditor
        hasBindingName
        valueDataBindingProvider={mockProvider}
        component={{} as IDataOverlayComponentInternal}
        onUpdateCallback={onUpdateCallbackMock}
      />,
    );

    expect(queryAllByTestId('binding-name-input').length).toBe(0);
    expect(container).toMatchSnapshot();
  });

  it('should render existing maps with bindings without binding name', async () => {
    const { container, queryAllByTestId } = render(
      <DataBindingMapEditor
        hasBindingName={false}
        skipFirstDivider
        valueDataBindingProvider={mockProvider}
        component={component}
        onUpdateCallback={onUpdateCallbackMock}
      />,
    );

    expect(queryAllByTestId('binding-name-input').length).toBe(0);
    expect(container).toMatchSnapshot();
  });
});
