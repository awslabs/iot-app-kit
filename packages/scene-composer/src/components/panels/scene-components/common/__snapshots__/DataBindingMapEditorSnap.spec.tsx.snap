// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataBindingMapEditor should render existing maps with bindings 1`] = `
.c0 {
  height: 1px;
  border-width: 0px;
  background-color: colorBorderDividerDefault;
}

.c1 {
  height: 8px;
}

<div>
  <div
    data-mocked="SpaceBetween"
  >
    <div
      data-mocked="Box"
    >
      <hr
        class="c0"
      />
      <div
        class="c1"
      />
      <div
        data-mocked="FormField"
        errortext="null"
        label="Binding Name"
      >
        <div
          data-mocked="Input"
          data-testid="binding-name-input"
          value="binding-1"
        />
      </div>
      <div
        data-mocked="Box"
        margin="{\\"vertical\\":\\"s\\"}"
      >
        <div
          data-mocked="SpaceBetween"
        >
          <div
            data-mocked="FormField"
            label="Entity Id"
          >
            <div
              data-mocked="Autosuggest"
              data-testid="select-entityId"
              options="[{\\"label\\":\\"label1\\",\\"value\\":\\"value1\\"},{\\"label\\":\\"label2\\",\\"value\\":\\"value2\\"}]"
              value="value1"
            />
          </div>
          <div
            data-mocked="FormField"
          >
            <div>
              <div
                data-mocked="Box"
              >
                Component Name
              </div>
            </div>
            <div
              data-mocked="Select"
              data-testid="value-data-binding-builder-select"
              options="[{\\"label\\":\\"label1\\",\\"value\\":\\"value1\\"},{\\"label\\":\\"label2\\",\\"value\\":\\"value2\\"}]"
              placeholder="Select an option"
              selectedoption="null"
            />
          </div>
          <div
            data-mocked="FormField"
          >
            <div>
              <div
                data-mocked="Box"
              >
                Property Name
              </div>
            </div>
            <div
              data-mocked="Select"
              data-testid="value-data-binding-builder-select"
              options="[{\\"label\\":\\"label1\\",\\"value\\":\\"value1\\"},{\\"label\\":\\"label2\\",\\"value\\":\\"value2\\"}]"
              placeholder="Select an option"
              selectedoption="null"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      data-mocked="Box"
    >
      <hr
        class="c0"
      />
      <div
        class="c1"
      />
      <div
        data-mocked="FormField"
        errortext="null"
        label="Binding Name"
      >
        <div
          data-mocked="Input"
          data-testid="binding-name-input"
          value="binding-2"
        />
      </div>
      <div
        data-mocked="Box"
        margin="{\\"vertical\\":\\"s\\"}"
      >
        <div
          data-mocked="SpaceBetween"
        >
          <div
            data-mocked="FormField"
            label="Entity Id"
          >
            <div
              data-mocked="Autosuggest"
              data-testid="select-entityId"
              options="[{\\"label\\":\\"label1\\",\\"value\\":\\"value1\\"},{\\"label\\":\\"label2\\",\\"value\\":\\"value2\\"}]"
              value="value1"
            />
          </div>
          <div
            data-mocked="FormField"
          >
            <div>
              <div
                data-mocked="Box"
              >
                Component Name
              </div>
            </div>
            <div
              data-mocked="Select"
              data-testid="value-data-binding-builder-select"
              options="[{\\"label\\":\\"label1\\",\\"value\\":\\"value1\\"},{\\"label\\":\\"label2\\",\\"value\\":\\"value2\\"}]"
              placeholder="Select an option"
              selectedoption="null"
            />
          </div>
          <div
            data-mocked="FormField"
          >
            <div>
              <div
                data-mocked="Box"
              >
                Property Name
              </div>
            </div>
            <div
              data-mocked="Select"
              data-testid="value-data-binding-builder-select"
              options="[{\\"label\\":\\"label1\\",\\"value\\":\\"value1\\"},{\\"label\\":\\"label2\\",\\"value\\":\\"value2\\"}]"
              placeholder="Select an option"
              selectedoption="null"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DataBindingMapEditor should render existing maps with bindings without binding name 1`] = `
.c1 {
  height: 1px;
  border-width: 0px;
  background-color: colorBorderDividerDefault;
}

.c0 {
  height: 8px;
}

<div>
  <div
    data-mocked="SpaceBetween"
  >
    <div
      data-mocked="Box"
    >
      <div
        class="c0"
      />
      <div
        data-mocked="Box"
        margin="{\\"vertical\\":\\"s\\"}"
      >
        <div
          data-mocked="SpaceBetween"
        >
          <div
            data-mocked="FormField"
            label="Entity Id"
          >
            <div
              data-mocked="Autosuggest"
              data-testid="select-entityId"
              options="[{\\"label\\":\\"label1\\",\\"value\\":\\"value1\\"},{\\"label\\":\\"label2\\",\\"value\\":\\"value2\\"}]"
              value="value1"
            />
          </div>
          <div
            data-mocked="FormField"
          >
            <div>
              <div
                data-mocked="Box"
              >
                Component Name
              </div>
            </div>
            <div
              data-mocked="Select"
              data-testid="value-data-binding-builder-select"
              options="[{\\"label\\":\\"label1\\",\\"value\\":\\"value1\\"},{\\"label\\":\\"label2\\",\\"value\\":\\"value2\\"}]"
              placeholder="Select an option"
              selectedoption="null"
            />
          </div>
          <div
            data-mocked="FormField"
          >
            <div>
              <div
                data-mocked="Box"
              >
                Property Name
              </div>
            </div>
            <div
              data-mocked="Select"
              data-testid="value-data-binding-builder-select"
              options="[{\\"label\\":\\"label1\\",\\"value\\":\\"value1\\"},{\\"label\\":\\"label2\\",\\"value\\":\\"value2\\"}]"
              placeholder="Select an option"
              selectedoption="null"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      data-mocked="Box"
    >
      <hr
        class="c1"
      />
      <div
        class="c0"
      />
      <div
        data-mocked="Box"
        margin="{\\"vertical\\":\\"s\\"}"
      >
        <div
          data-mocked="SpaceBetween"
        >
          <div
            data-mocked="FormField"
            label="Entity Id"
          >
            <div
              data-mocked="Autosuggest"
              data-testid="select-entityId"
              options="[{\\"label\\":\\"label1\\",\\"value\\":\\"value1\\"},{\\"label\\":\\"label2\\",\\"value\\":\\"value2\\"}]"
              value="value1"
            />
          </div>
          <div
            data-mocked="FormField"
          >
            <div>
              <div
                data-mocked="Box"
              >
                Component Name
              </div>
            </div>
            <div
              data-mocked="Select"
              data-testid="value-data-binding-builder-select"
              options="[{\\"label\\":\\"label1\\",\\"value\\":\\"value1\\"},{\\"label\\":\\"label2\\",\\"value\\":\\"value2\\"}]"
              placeholder="Select an option"
              selectedoption="null"
            />
          </div>
          <div
            data-mocked="FormField"
          >
            <div>
              <div
                data-mocked="Box"
              >
                Property Name
              </div>
            </div>
            <div
              data-mocked="Select"
              data-testid="value-data-binding-builder-select"
              options="[{\\"label\\":\\"label1\\",\\"value\\":\\"value1\\"},{\\"label\\":\\"label2\\",\\"value\\":\\"value2\\"}]"
              placeholder="Select an option"
              selectedoption="null"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DataBindingMapEditor should render with no bindings 1`] = `
<div>
  <div
    data-mocked="SpaceBetween"
  />
</div>
`;
